### PR TITLE
Fix avatar generator centering

### LINKED ISSUE #8841: Avatar replacement character for accentuated letter is not centered
NC 13.0.1 includes https://github.com/nextcloud/server/pull/8468 that does correct https://github.com/nextcloud/server/issues/8451 but the avatar is not centered correctly. The accent should not be taken into account to position the letter in the circle.

For example the bottom bar of "É" should be on the same horizontal level as "E". And it's not, because the accent is taken into account to position the letter in the circle. Unexpected result: "É" is positionned too low.

image

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix #8841 

@nextcloud/designers @biva 

| Before | After |
|:-------:|:------:|
| avatar 64 5 | avatar 64 |
| avatar 64 4 | avatar 64 1 |
| avatar 64 3 | avatar 64 2 |
| avatar 64 6 | avatar 64 7 |

### CODE REVIEW

File: lib/private/Avatar.php [hunk 1]
```diff
@@ -291,6 +288,35 @@ private function generateAvatar($userDisplayName, $size) {
...
+		$yi = imagesy($image);
+
+		// bounding box
+		$box = imagettfbbox($size, $angle, $font, $text);
+
+		// imagettfbbox can return negative int
+		$xr = abs(max($box[2], $box[4]));
+		$yr = abs(max($box[5], $box[7]));
+
+		// calculate bottom left plcaement
```
Comment (MorrisJobke) [inline lib/private/Avatar.php]: Typo ;)

**Comment by skjnldsv** [comment]
> I don't see a difference here (1 and 3 were created on master, 2 and 4 were created on this branch)

Are you sure the cache was cleared and stuff, because if you look at my first post, you'll see we really have a difference :)
Nonetheless, the centering still ignores the accent, so the letters are still a bit too low, I'll check what I can do

**Review by danxuliu — COMMENTED** [review]
I can see a difference between avatars generated in _master_ branch and avatars generated in this branch. In general they look better now :-) (although for some specific characters they were better before, like for _A_).

Unfortunately I do not think that it will be possible to get perfect centering with small sizes if sub-pixel accuracy is not provided by the underlying libraries :-( (@skjnldsv you cheated a little with the examples in the first post, as with 64x64px the avatar can be better centered than with 32x32px :-P ). Anyway, this looks good enough ;-)

**Comment by skjnldsv** [comment]
> (@skjnldsv you cheated a little with the examples in the first post, as with 64x64px the avatar can be better centered than with 32x32px :-P ). Anyway, this looks good enough ;-)

Damn, busted! :D 
Let put this pr on hold, and let me see if I can find a way of properly generate them :)
Especially if I can respect the font baseline instead of box ;)

**Comment by rullzer** [comment]
Maybe just generate the avatar at 128x128 and scale it down :P?

**Comment by skjnldsv** [comment]
@rullzer actually, I think that's how I'm heading. 
Generate svg file and convert to png :)

**Comment by danxuliu** [comment]
Scaling the text down will probably cause it to be too blurry :-( Even with a SVG file you would need to somehow align the character to the grid so it looks crisp once rasterized (font hinting), although maybe that is automatically done by the library, no idea :-)

**Comment by skjnldsv** [comment]
Looks pretty good to me :D avatar

**Comment by danxuliu** [comment]
>  Looks pretty good to me :D

Indeed; I was expecting it to be much worse. Fine by me then ;-)

**Comment by MorrisJobke** [comment]
CI is fixed on master.

**Comment by skjnldsv** [comment]
As a reference, here are our solution and their issues
1. Convert svg to png in server:
   - we need Imagemagick to convert svg, gd does not supports it
   - we can use phantomjs to trick the conversion, but  not very convenient and EOL has been announced for phantomjs
2. Pre-generate all avatar
   - We can't generate all non-latin chars. It would be fine  for latin only, but we need to keep support for japanese, idian... 
3. Serve svg
   - Web supports it perfectly
   - Lighter than generating png
   - IOS & Android have some issues with it. Currently @Ivansss and @tobiasKaminsky are looking at our option this week and we'll see what we can do later
4. Serve png AND svg: If svg supported: serve svg if not, serve png
   - Very bad ux since we will have good avatar and 'bad' ones

**Comment by tobiasKaminsky** [comment]
I just took a look into our code and we already using svg rendering on activity/notification list.
As the images are cached locally, I hope that there is no big speed drawback.

All in all I would like to give svg avatars a try. 
Currently we are not using avatars that often.

**Comment by Ivansss** [comment]
iOS is not SVG-friendly (SVG images are not supported by default).
We could use third-party libraries to render SVGs, but the performance/user experience would be definitely affected.
So I would go with the 4th option or the 1st one if possible.
Since the server generate and serve the avatars, I think it should also convert them to png.

**Comment by skjnldsv** [comment]
@Ivansss how does the avatar looks on ios right now? Are the letters shifted to the right or the left? Is it centered?

**Comment by rullzer** [comment]
We won't do the 1st option. We disabled the required dependency on imagemagik for a reason ;)

**Comment by Ivansss** [comment]
Sorry @skjnldsv , I totally missed your message.
They look like this at the moment.
img_0148
So a bit shifted to the right? Depends on the character I would say.
Could the 1st option be done using another library instead of imagemagick?

**Comment by skjnldsv** [comment]
@Ivansss no problem :)
Can you try with the "É" letter?

**Comment by Ivansss** [comment]
img_e0149

**Comment by biva** [comment]
:(
É should be at the exact same place as E. We shouldn't take the accent into account

**Comment by skjnldsv** [comment]
@biva that's the whole point of this pr ^^
@Ivansss thanks, we will introduce the fallback to png, at least the letter is better centered now, but the accent is still taken in consideration in the png generation :/

**Comment by biva** [comment]
@skjnldsv sorry ! I thought it was the new version. I let you work guys, sorry again.

**Comment by skjnldsv** [comment]
@biva no need to apologise :hugs: :)

This is the new version for the png generation. We will nonetheless introduce the svg avatar as well as primary avatar for supported systems (all except ios ^^) which have the letters all properly aligned ;)

see https://github.com/nextcloud/server/pull/9197#issuecomment-381619266

**Comment by MorrisJobke** [comment]
@skjnldsv Is this ready for review?

**Comment by skjnldsv** [comment]
@MorrisJobke Yes, I was waiting for the tests :)
Let's go! :writing_hand:

**Comment by MorrisJobke** [comment]
I don't have "Open Sans" installed on my system and then the avatar looks like this:

bildschirmfoto 2018-05-07 um 16 27 16

**Comment by skjnldsv** [comment]
@MorrisJobke damn it, that's something we did not think of :/

**Comment by MorrisJobke** [comment]
> @MorrisJobke damn it, that's something we did not think of :/

Render it on the server side and then generate a PNG out of it? Then also the iOS app will work with this avatar without problems.

**Comment by pixelipo** [comment]
Those are rendered as SVG? CSS would solve both the font issue and the bottom alignment. What am I missing here?

**Comment by skjnldsv** [comment]
@pixelipo ! Welcome back :hugs: :)
Yes, that's why we're switching to svg avatars, because the css allow us to properly align the letters from the baseline.
The only issue is the font. A svg loaded as img, does not take the page css font-face in account :/

**Comment by skjnldsv** [comment]
Because it's being served as a image/xml+svg for our other devices. And it's easier to just use the userid in url `http://xxx/avatar/userid/size` in an `img` tag than use javascript to load data and stuff, or implement if on our page directly.

This way we only have one get point for th avatar and only one location to edit if we want to change the process (like we're doing right now) instead of everywhere where the div would have been use :)

**Comment by skjnldsv** [comment]
@MorrisJobke can you try again? :)

**Comment by MorrisJobke** [comment]
Still doesn't show up properly:

bildschirmfoto 2018-05-23 um 14 02 53

**Comment by skjnldsv** [comment]
@nextcloud/designers Okay
I changed our way of doing. If imagick is installed, the svg avatar will be used and converted to png.
If not, the old (but improved) png gd avatar will be used. centered horizontally but can have some glitches with accents or custom chars.

It's better than nothing though :)

**Comment by MorrisJobke** [comment]
Works now, but the initial goal of centering letters with special signs above or below them is still not reached:

bildschirmfoto 2018-05-28 um 15 44 17

I'm fine with getting this in due to the color re-calculation nevertheless.

**Review by MorrisJobke — APPROVED** [review]

**Comment by skjnldsv** [comment]
@MorrisJobke 

> I changed our way of doing. If imagick is installed, the svg avatar will be used and converted to png.
If not, the old (but improved) png gd avatar will be used. centered horizontally but can have some glitches with accents or custom chars.

You probably don't have imagick set up :)

**Comment by MorrisJobke** [comment]
> You probably don't have imagick set up :)

It says that it is installed:

bildschirmfoto 2018-05-28 um 15 53 59

**Comment by skjnldsv** [comment]
@MorrisJobke maybe a cache issue?
Try getting a new avatar size?
`https://yourdomain.com/avatar/Test/100`

**Comment by skjnldsv** [comment]
Any news?

**Review by MorrisJobke — APPROVED** [review]

**Comment by MorrisJobke** [comment]
Restarted CI - then this can go in

**Comment by skjnldsv** [comment]
@MorrisJobke Here we go! :see_no_evil:

**Pull request merged**